library/BerlinPipeline/src/Juvix/BerlinPipeline/CircularList.hs

```diff
@@ -1,6 +1,48 @@
 module Juvix.BerlinPipeline.CircularList where
 
 import Juvix.Library
+import qualified Juvix.Library.NameSymbol as NameSymbol
 
-data T a
+data T a = T (RecList (CircSchema a))
   deriving (Show, Eq)
+
+instance Semigroup (T a) where
+  T l <> T r = T (l <> r)
+
+data CircSchema a
+  = CircSchema [a]
+  | NonCircSchema a
+  deriving (Show, Eq)
+
+data RecList a
+  = Rec [RecList a]
+  | Anu a
+  deriving (Show, Eq)
+
+firstNested :: T a -> Maybe (CircSchema a)
+firstNested (T (Anu a)) = Just a
+firstNested (T (Rec [])) = Nothing
+firstNested (T (Rec (x : _xs))) = firstNested (T x)
+
+removeFirstNested :: T a -> T a
+removeFirstNested (T (Anu a)) = T (Rec [])
+removeFirstNested (T (Rec [])) = T (Rec [])
+removeFirstNested (T (Rec (x : xs))) =
+  case removeFirstNested (T x) of
+    T (Rec []) -> T (Rec xs)
+    T otherwise -> T (Rec (otherwise : xs))
```

Comment:
We should probably move the calls that match into `Rec` into the `Reclist` module itself, as Reclist should have these same functions, and would require less wrapping and unwrapping